RecoLocalCalo/HcalRecAlgos/src/MahiGPU.cu

```diff
@@ -0,0 +1,1901 @@
+        printf("acc25nsVec and diff25nsItvlVec for recoPulseShapeId = %u\n",
+            recoPulseShapeId);
+        for (int i=0; i<256; i++) {
+            printf("acc25nsVec(%d) = %f diff25nsItvlVec(%d) = %f\n",
+                i, acc25nsVec[i], i, diff25nsItvlVec[i]);
+        }
+        printf("accVarLenIdxZEROVec and accVarLenIdxMinusOneVec\n");
+        for (int i=0; i<25; i++) {
+            printf("accVarLenIdxZEROVec(%d) = %f accVarLenIdxMinusOneVec(%d) = %f\n",
+                i, accVarLenIdxZeroVec[i], i, accVarLenIdxMinusOneVec[i]);
+        }
+        printf("diffVarItvlIdxZEROVec and diffVarItvlIdxMinusOneVec\n");
+        for (int i=0; i<25; i++) {
+            printf("diffVarItvlIdxZEROVec(%d) = %f diffVarItvlIdxMinusOneVec(%d) = %f\n",
+                i, diffVarItvlIdxZeroVec[i], i, diffVarItvlIdxMinusOneVec[i]);
+        }
+    }
+#endif
+
+    auto t0 = meanTime;
+    if (applyTimeSlew) {
+        if (amplitude <= 1.0f) 
+            t0 += compute_time_slew_delay(1.0, 
+                tzeroTimeSlew, slopeTimeSlew, tmaxTimeSlew);
+        else 
+            t0 += compute_time_slew_delay(amplitude,
+                tzeroTimeSlew, slopeTimeSlew, tmaxTimeSlew);
+    }
+    auto const t0m = -deltaT + t0;
+    auto const t0p = deltaT + t0;
+
+#ifdef HCAL_MAHI_GPUDEBUG
+    if (sample == 0 && ipulse == 0) {
+        printf("time values: %f %f %f\n", t0, t0m, t0p);
+    }
+
+    if (sample == 0 && ipulse == 0) {
+        for (int i=0; i<10; i++) {
+            auto const value = compute_pulse_shape_value(t0, i, 0,
+                acc25nsVec, diff25nsItvlVec,
+                accVarLenIdxMinusOneVec, diffVarItvlIdxMinusOneVec,
+                accVarLenIdxZeroVec, diffVarItvlIdxZeroVec);
+            printf("pulse(%d) = %f\n", i, value);
+        }
+        printf("\n");
+        for (int i=0; i<10; i++) {
+            auto const value = compute_pulse_shape_value(t0p, i, 0,
+                acc25nsVec, diff25nsItvlVec,
+                accVarLenIdxMinusOneVec, diffVarItvlIdxMinusOneVec,
+                accVarLenIdxZeroVec, diffVarItvlIdxZeroVec);
+            printf("pulseP(%d) = %f\n", i, value);
+        }
+        printf("\n");
+        for (int i=0; i<10; i++) {
+            auto const value = compute_pulse_shape_value(t0m, i, 0,
+                acc25nsVec, diff25nsItvlVec,
+                accVarLenIdxMinusOneVec, diffVarItvlIdxMinusOneVec,
+                accVarLenIdxZeroVec, diffVarItvlIdxZeroVec);
+            printf("pulseM(%d) = %f\n", i, value);
+        }
+    }
+#endif
+
+    // FIXME: shift should be treated properly, 
+    // here assume 8 time slices and 8 samples
+    auto const shift = 4 - soi; // as in cpu version!
+//    auto const offset = ipulse - soi;
+//    auto const idx = sample - offset;
+    auto const idx = sample - pulseOffset;
+    auto const value = idx>=0 && idx<nsamples
+        ? compute_pulse_shape_value(t0, idx, shift,
```

Comment:
For each rechit, is this an array of templates (we need one for each BX) or you approximated all the template(covmatrices) to be the same for each BX?

Reply:
everything in there is done according to PulseShapeFunctor.cc. a related conditions product to check is HcalRecoParamsWithPulseShapesGPU.cc/.h

Reply:
uhm, indeed that is what is seems.

but PulseShapeFunctor.cc just give you A template, 
but then inside nnls fit works with an array of template and covariance matrices.  
Since we put 8 pulses to fit 8 digis, each pulse carry a templatevector, a derivative vector and a covariance matric.

Read carefully the function doFit() in the cms-sw master

Reply:
i suggest you check the code carefully in the way it is written and the logic of the original. it's validated and reproduces cpu results, using hcal_mahi_patatrack_validation branch, almost 1 to 1. 

 - First of, you do not need an array of matrices
 - second, u need just 3 matrices, that is what `kernel_prep_pulseMatrices_sameNumberOfSamples` does - prepares your templates, except u do not need array of those, but rather just 3 of them
 - third, in update covariance matrix (in kernel minimize), you can __compute (sort of)__ your array of template matrices, rather than storing all of them.

Reply:
the reason you have an array of those is cause then it's much easier to use `eigen`

Reply:
The point is not how you organized them in memory, the question is: for the covariance matrix do you have 8*64 different numbers for each rechit ? same for the template central values you should have a total of 8*8 different numbers (here in particular position you need to have zero).
It's not clear to me how you have done here.
Thes many numbers comes for the time-slew modeling otherwise we could have calculated the templates only twice per run (once for HPD and once for siPM).

Reply:
- you do not need 8 x 64, you can do with 3 x 64. please check the logic and symmetry of these matrices that you use. In short, for each pulse you need 3 x 8 values (central, plus, minus). Moreover, if u guys did not use time slew, these matrices would be completely static. Again, check how you build them and what is the input... obviously, given there is a time slew computation which is adc per ts dependent, you get variations for each channel. but i'm just showing an example here of what can be precomputed and not redone for each channel. for cpu most likely this is not dramatic, but still. for gpu that counts...   
 - your PulseShapeFunctor.cc gets reinitialized for every channel. The input is pulse shape, pulse shape is not coming in for each channel but rather per qieType, __please trace how things go back to conditions!__ You can see how this is treated starting from __what sits in conditions__ in `HcalRecoParamsWithPulseShapesGPU.cc`. What you guys recompute for every channel, i just precompute for each qie type.
 - then in `kernel_prep_pulseMatrices_sameNumberOfSamples` i build the same 8 x 64 that you guys have, except that i use 3 * 8 * 8, cause you got matrcies that are symmetric. moreover your covariance matrix that you compute is also symmetric -> less loads needed/etc... for small matrices this is important. 
- The idea is that you guys keep around large number of matrices which are not needed at all, this is simplified down to what actually is needed and gets used. You operate on symmetric matrices in there in most cases.